client/publisher.go

```diff
@@ -0,0 +1,105 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/conduitio/conduit-connector-gcp-pubsub/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+// A PublisherInterface represents a publisher interface.
+type PublisherInterface interface {
+	Publish(context.Context, sdk.Record, sdk.AckFunc)
+	Stop() error
+}
+
+// A Publisher represents a publisher struct with a GCP Pub/Sub client.
+type Publisher struct {
+	*pubSub
+
+	topic     *pubsub.Topic
+	resultsCh chan *result
+	ctxCancel context.CancelFunc
+}
+
+type result struct {
+	publishResult *pubsub.PublishResult
+	ackFunc       sdk.AckFunc
+}
+
+// NewPublisher initializes a new publisher client.
+func NewPublisher(ctx context.Context, cfg config.Destination) (PublisherInterface, error) {
+	ps, err := newClient(ctx, cfg.General)
+	if err != nil {
+		return nil, fmt.Errorf("new pubsub client: %w", err)
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+
+	publisher := &Publisher{
+		pubSub:    ps,
+		topic:     ps.client.Topic(cfg.TopicID),
+		resultsCh: make(chan *result, cfg.BatchSize),
+		ctxCancel: cancel,
+	}
+
+	publisher.topic.PublishSettings.CountThreshold = cfg.BatchSize
+	publisher.topic.PublishSettings.DelayThreshold = cfg.BatchDelay
+
+	go func() {
+		for {
+			select {
+			case <-cctx.Done():
+				return
+			case msg := <-publisher.resultsCh:
+				<-msg.publishResult.Ready()
```

Comment:
We need to wait until the message is ready, and only after that - call `Get` to ensure the message was successfully published.
If you call `Get` earlier, the message will be sent immediately, not in batches.

Reply:
Okay, I was asking because I saw the [Publish with batching settings code sample](https://cloud.google.com/pubsub/docs/samples/pubsub-publisher-batch-settings#code-sample) provided in the GCP Pub/Sub docs, and I couldn't find any mention of the `Ready` method there.

Reply:
I see.

As you can see in the example, the first action is `Publish` some messages, and only after that - `Get`.
In our case, we should do `Publish` and `Get` (to ensure the message was successfully published) for each record.
But if call both methods each time - it wouldn't be as an async writer.

Reply:
It seems to me that the `Get` method is doing pretty much the [same job under the hood](https://github.com/googleapis/google-cloud-go/blob/v0.102.1/internal/pubsub/publish.go#L31). And I really didn't notice any difference when deleted the `Ready` call from your code.

Reply:
Ok, agree with you.
Deleted it.